camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/EffectorsYamlTest.java

```diff
@@ -18,45 +18,168 @@
+            "      $brooklyn:entity(\"entity1\").effector(\"identityEffector\", \"hello\")"
+        );
+        TestEntity testEntity = (TestEntity)Iterables.getOnlyElement(app.getChildren());
+        Assert.assertEquals(testEntity.getConfig(TestEntity.CONF_NAME), "hello");
+        assertCallHistory(testEntity, "start", "identityEffector");
+    }
+
+    @Test
+    public void testOwnEffectorWithReturn() throws Exception {
+
+        Entity app = createAndStartApplication(
+            "services:",
+            "- type: " + TestEntity.class.getName(),
+            "  brooklyn.config:",
+            "    test.confName: ",
+            "      $brooklyn:effector(\"identityEffector\", \"my own effector\")"
+        );
+        TestEntity testEntity = (TestEntity)Iterables.getOnlyElement(app.getChildren());
+        Assert.assertEquals(testEntity.getConfig(TestEntity.CONF_NAME), "my own effector");
+        assertCallHistory(testEntity, "start", "identityEffector");
+    }
+
+    private void assertCallHistory(TestEntity testEntity, String... expectedCalls) {
+        List<String> callHistory = testEntity.getCallHistory();
+        Assert.assertEquals(callHistory.size(), expectedCalls.length, "history = " + callHistory);
+        int c = 0;
+        for (String expected : expectedCalls) {
+            Assert.assertEquals(callHistory.get(c++), expected, "history = " + callHistory);
         }
-        Assert.assertNotNull(start1);
-        Assert.assertNotNull(start2);
-        EntityAsserts.assertAttributeEquals(start1, Startable.SERVICE_UP, false);
-        EntityAsserts.assertAttributeEquals(start2, Startable.SERVICE_UP, true);
     }
-    
 
+    @Test
+    public void testEffectorCalledOncePerConfigKey() throws Exception {
+
+        Entity app = createAndStartApplication(
+            "services:",
+            "- type: " + TestEntity.class.getName(),
+            "  id: entity1",
+            "  brooklyn.config:",
+            "    test.confName: ",
+            "      $brooklyn:effector(\"sequenceEffector\")",
+            "    test.confObject: ",
+            "      $brooklyn:effector(\"sequenceEffector\")"
+        );
+        TestEntity testEntity = (TestEntity)Iterables.getOnlyElement(app.getChildren());
+
+        List<String> callHistory = testEntity.getCallHistory();
+        Assert.assertFalse(callHistory.contains("myEffector"), "history = " + callHistory);
+
+        final String firstGetConfig = testEntity.getConfig(TestEntity.CONF_NAME);
+        Assert.assertEquals(firstGetConfig, "1");
+        final String secondGetConfig = testEntity.getConfig(TestEntity.CONF_NAME);
+        Assert.assertEquals(secondGetConfig, "1");
+        Assert.assertEquals(testEntity.getConfig(TestEntity.CONF_OBJECT), Integer.valueOf(2));
+        assertCallHistory(testEntity, "start", "sequenceEffector", "sequenceEffector");
+    }
+
+    @Test(groups = "Integration")
+    public void testSshCommandSensorWithEffectorInEnv() throws Exception {
```

Comment:
Personal preference for keeping the integration tests like this in a separate class. We can also use a `RecordingSshTool` so that it avoids any ssh commands, and thus keep it as a unit test.